docs/resiliency-checklist.md

```diff
@@ -83,12 +83,24 @@ It is preferable to scale horizontally than vertically. Use smaller Request and
 
 Make use of the [Quality of Service](https://docs.openshift.com/container-platform/3.11/dev_guide/compute_resources.html#quality-of-service-tiers) options available to ensure your pods are correctly resourced and treated appropriately by the scheduler.
 
-If your pod has its Requests and Limits set to `0` it will run BestEffort and simply use whatever spare capacity is available on its node. This can be good for things like batch jobs that want to go as fast as they can, but don't care about being slowed down from time to time. It will also be the first to be evicted from an overloaded node. Assigning `0` as a request or limit must be done through the CLI or directly in the manifest. 
+If your pod has its Requests and Limits set to `0` it will run BestEffort and simply use whatever spare capacity is available on its node. This can be good for things like batch jobs that want to go as fast as they can, but don't care about being slowed down from time to time. It will also be the first to be evicted from an overloaded node. Assigning `0` as a request or limit must be done through the CLI or directly in the manifest.
 
 If your pod has Limits higher than its Requests it will run Burstable and get at minimum the Requested amount and by burst up to its Limit depending on how busy the node is with other pods.
 
 If your pod has the Requests and Limits set to the same value then it will run Guaranteed QoS class. It will have preferential access to compute resources and will be the last to be evicted if a node should become overloaded. This is best for apps in the `prod` namespace as they need the best uptime.
 
+### A Well Behaved App
+
+It is important to ensure your app is well behaved and doesn't impede the cluster's operators work or impact the cluster's ability to heal.
+
+Review the Kubernetes docs on [pod terminations](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) to understand the full sequence of events. Ensure that your pods will gracefully exit. This means not setting the `terminationGracePeriod` too high, and ensuring your containers command properly listens for the `SIGTERM` and responds.
```

Comment:
Completely agree that ensuring containers have been built so they properly respond to the `SIGTERM` is critical. Where we can't do that for whatever reason, a `PreStop` hook can be used.

For a "stateless" webserver, a graceful shutdown of process often involves the following steps:
- An application must wait a few seconds, then stop accepting new connections;
- An application must wait until all requests are complete and close all idle keepalive connections;
- It must terminate its own process.

If we don't wait for shutdown we cause 503 errors by disrupting the stateful tcp connection. Since our web application itself does not hold any state, only the tcp connections clients have open to it are stateful, the harm caused by force-killing a misbehaving process is limited to temporary unavailability of the service. And the magic of k8s is that our kube-proxy acts as a layer 4 load balancer for our service and will terminate the zombie tcp socket in accordance with configured [tcp keepalive period](https://github.com/golang/go/blob/release-branch.go1.9/src/net/tcpsockopt_unix.go#L15-L25). If memory serves, tcp keepalive by default needs 9 unsuccessful proves to kill connection this means it need in total 4m30sec to kill connection if it was idle for 30sec. So failure to gracefully terminate tcp sockets to a pod might cause a larger outage than naively anticipated, but will eventually self-heal at the service level thanks kube-proxy. Most importantly, no permanent damage from the outage will occur: no data will be lost since the application itself is not stateful, even though it may be transporting state instruction messages over a stateful tcp socket.

For this kind of stateless application, setting a `terminationGracePeriod` too high delays the amount of time between attempting a node drain and discovering an application is not responding to `SIGTERM`, whether that's due to misconfiguration, resource starvation, an undefined state, or whatever reason. I fully support force killing these kind of processes after a reasonable grace.